File: 上新/modules/price_review.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time
from datetime import datetime
import os
from .price_mapper import PriceMapper

class PriceReview:
    def __init__(self, driver, wait, debug=True):
        """
        初始化价格审核类
        :param driver: WebDriver实例
        :param wait: WebDriverWait实例
        :param debug: 是否开启调试模式
        """
        self.driver = driver
        self.wait = wait
        self.debug = debug
        self.price_mapper = PriceMapper(debug)
        
        # 创建日志目录
        self.log_dir = "logs"
        if not os.path.exists(self.log_dir):
            os.makedirs(self.log_dir)
        
        # 创建日志文件（使用日期作为文件名）
        self.log_file = os.path.join(self.log_dir, f"price_review_{datetime.now().strftime('%Y%m%d')}.log")
        
        # 如果文件不存在，添加表头
        if not os.path.exists(self.log_file):
            with open(self.log_file, "w", encoding="utf-8") as f:
                f.write("=" * 80 + "\n")
                f.write(f"价格审核日志 - {datetime.now().strftime('%Y-%m-%d')}\n")
                f.write("=" * 80 + "\n\n")

    def log_record(self, sku, spu, min_price, new_price, action):
        """
        记录价格审核日志
        :param sku: 货号
        :param spu: SPU信息
        :param min_price: 最低核价
        :param new_price: 新申报价格
        :param action: 执行的操作
        """
        try:
            timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            log_entry = f"[{timestamp}] 货号: {sku}, SPU: {spu}, 最低核价: {min_price}, 新申报价格: {new_price}, 操作: {action}\n"
            
            with open(self.log_file, "a", encoding="utf-8") as f:
                f.write(log_entry)
                
            if self.debug:
                print(f"已记录日志: {log_entry.strip()}")
                
        except Exception as e:
            print(f"记录日志失败: {str(e)}")

    def open_price_review_page(self):
        """
        打开价格审核页面
        """
        try:
            if self.debug:
                print("正在打开上新主页面...")

            # 打开上新主页面
            self.driver.get("https://seller.kuajingmaihuo.com/main/product/seller-select")
            
            # 等待页面加载
            time.sleep(3)
            
            if self.debug:
                print("页面加载完成")

        except Exception as e:
            print(f"打开价格审核页面失败: {str(e)}")
            raise

    def click_price_pending_button(self):
        """
        点击价格待确认按钮
        """
        try:
            if self.debug:
                print("正在查找价格待确认按钮...")

            # 使用CSS选择器定位按钮
            button = self.wait.until(
                EC.element_to_be_clickable((By.CSS_SELECTOR, '#root > div > div > div > div:nth-child(1) > div:nth-child(1) > div:nth-child(2) > div:nth-child(2) > div:nth-child(6)'))
            )
            
            # 获取按钮文本（包含商品数量）
            button_text = button.text
            if self.debug:
                print(f"找到价格待确认按钮: {button_text}")

            # 点击按钮
            button.click()
            
            if self.debug:
                print("成功点击价格待确认按钮")

        except Exception as e:
            print(f"点击价格待确认按钮失败: {str(e)}")
            raise

    def get_product_list(self):
        """
        获取商品列表
        :return: 商品列表元素
        """
        try:
            if self.debug:
                print("正在获取商品列表...")

            # 等待商品列表加载
            product_list = self.wait.until(
                EC.presence_of_element_located((By.CSS_SELECTOR, '#root > div > div > div > div.TB_outerWrapper_5-117-0.TB_bordered_5-117-0.TB_notTreeStriped_5-117-0 > div.TB_inner_5-117-0 > div > div.TB_body_5-117-0 > div > div > table > tbody'))
            )
            
            if self.debug:
                print("商品列表加载完成")
            
            return product_list

        except Exception as e:
            print(f"获取商品列表失败: {str(e)}")
            raise

    def get_product_info(self, row_element):
        """
        获取单个商品的信息
        :param row_element: 商品行元素
        :return: 商品信息字典
        """
        try:
            # 获取SPU信息
            spu_element = row_element.find_element(By.CSS_SELECTOR, 'div.product-info_name__pXkCK')
            spu_info = spu_element.text if spu_element else "未找到SPU信息"

            # 获取货号
            sku_element = row_element.find_element(By.CSS_SELECTOR, 'div.use-columns_infoItem__iPDoh.use-columns_light__sKFcQ')
            sku_number = sku_element.text if sku_element else "未找到货号"

            # 获取价格确认按钮
            price_button = row_element.find_element(By.CSS_SELECTOR, 'button.BTN_outerWrapper_5-111-0')
            
            # 获取价格信息
            code, min_price = self.price_mapper.get_price_info(sku_number)
            
            if self.debug:
                print(f"\n商品信息:")
                print(f"SPU: {spu_info}")
                print(f"货号: {sku_number}")
                print(f"识别码: {code}")
                print(f"最低核价: {min_price}")
                print(f"找到价格确认按钮: {price_button.text}")

            return {
                'spu': spu_info,
                'sku': sku_number,
                'code': code,
                'min_price': min_price,
                'price_button': price_button
            }

        except Exception as e:
            print(f"获取商品信息失败: {str(e)}")
            raise

    def handle_price_confirmation(self, min_price, sku, spu):
        """
        处理价格确认弹窗
        :param min_price: 最低核价
        :param sku: 货号
        :param spu: SPU信息
        :return: 是否成功处理
        """
        try:
            if self.debug:
                print("\n开始处理价格确认弹窗...")

            # 等待弹窗出现
            modal = self.wait.until(
                EC.presence_of_element_located((By.XPATH, "/html/body/div[7]/div/div/div[1]"))
            )
            
            # 获取新申报价格
            new_price_element = self.wait.until(
                EC.presence_of_element_located((By.XPATH, "/html/body/div[7]/div/div/div[1]/div[2]/form/div/div/div/table/tbody/tr/td[5]/div/div/div/div/span/span[2]"))
            )
            new_price = float(new_price_element.text)
            
            if self.debug:
                print(f"最低核价: {min_price}")
                print(f"新申报价格: {new_price}")

            # 比较价格
            if new_price > min_price:
                if self.debug:
                    print("新申报价格大于最低核价，直接确认")

                
                # 等待用户观察
                if self.debug:
                    print("等待5秒，请观察确认结果...")
                time.sleep(5)
                
                # 点击确认提交按钮
                confirm_button = self.wait.until(
                    EC.element_to_be_clickable((By.XPATH, "/html/body/div[7]/div/div/div[1]/div[3]/div[2]/button[1]"))
                )
                confirm_button.click()
                
                # 记录日志
                self.log_record(sku, spu, min_price, new_price, "确认提交")
                
                
            else:
                if self.debug:
                    print("新申报价格小于等于最低核价，选择放弃调整报价")
                
                # 点击下拉框
                dropdown = self.wait.until(
                    EC.element_to_be_clickable((By.XPATH, "//*[@id='operatorType']/div/div/div/div/div/div/div/div/div/div"))
                )
                dropdown.click()
                
                # 选择"放弃调整报价"选项
                abandon_option = self.wait.until(
                    EC.element_to_be_clickable((By.XPATH, "/html/body/div[8]/div/div/div/div/ul/li[3]"))
                )
                abandon_option.click()

                # 等待用户观察
                if self.debug:
                    print("等待5秒，请观察确认结果...")
                time.sleep(5)
                
                # 点击确认提交按钮
                confirm_button = self.wait.until(
                    EC.element_to_be_clickable((By.XPATH, "/html/body/div[7]/div/div/div[1]/div[3]/div[2]/button[1]"))
                )
                confirm_button.click()
                
                # 记录日志
                self.log_record(sku, spu, min_price, new_price, "放弃调整报价")
                

            # 等待弹窗消失
            time.sleep(2)
            
            if self.debug:
                print("价格确认处理完成")
            
            return True

        except Exception as e:
            print(f"处理价格确认弹窗失败: {str(e)}")
            # 记录错误日志
            self.log_record(sku, spu, min_price, "获取失败", f"处理失败: {str(e)}")
            return False

    def process_product_list(self):
        """
        处理商品列表
        """
        try:
            # 获取商品列表
            product_list = self.get_product_list()
            
            # 获取所有商品行
            rows = product_list.find_elements(By.TAG_NAME, "tr")
            
            if self.debug:
                print(f"\n找到 {len(rows)} 个商品")

            # 处理每一行商品
            for index, row in enumerate(rows, 1):
                if self.debug:
                    print(f"\n处理第 {index} 个商品:")
                
                # 获取商品信息
                product_info = self.get_product_info(row)
                
                # 点击价格确认按钮
                product_info['price_button'].click()
                
                # 处理价格确认弹窗
                if product_info['min_price'] is not None:
                    self.handle_price_confirmation(
                        product_info['min_price'],
                        product_info['sku'],
                        product_info['spu']
                    )
                
                if self.debug:
                    print(f"已处理第 {index} 个商品")
                
                # 等待一下，避免操作太快
                time.sleep(1)

        except Exception as e:
            print(f"处理商品列表失败: {str(e)}")
            raise

    def start_review(self):
        """
        开始价格审核流程
        """
        try:
            # 打开价格审核页面
            self.open_price_review_page()
            
            # 点击价格待确认按钮
            self.click_price_pending_button()
            
            # 等待页面加载
            time.sleep(3)
            
            # 处理商品列表
            self.process_product_list()
            
            if self.debug:
                print("价格审核流程初始化完成")

        except Exception as e:
            print(f"价格审核流程初始化失败: {str(e)}")
            raise 

File: ProductListingWorkflow/src/services/price-review.ts
import { Page, BrowserContext } from '@playwright/test';
import { logger } from '../utils/logger';
import { PriceMapper } from './price-mapper';
import * as fs from 'fs';
import * as path from 'path';

export class PriceReview {
    private page: Page;
    private priceMapper: PriceMapper;
    private logDir: string;
    private logFile: string;
    private errorCount: number = 0;
    private readonly MAX_ERROR_COUNT: number = 5;

    constructor(context: BrowserContext, debug: boolean = true) {
        this.page = context.pages()[0];
        this.priceMapper = new PriceMapper(debug);
        
        // 创建日志目录
        this.logDir = "logs";
        if (!fs.existsSync(this.logDir)) {
            fs.mkdirSync(this.logDir);
        }
        
        // 创建日志文件（使用日期作为文件名）
        const date = new Date().toISOString().split('T')[0].replace(/-/g, '');
        this.logFile = path.join(this.logDir, `price_review_${date}.log`);
        
        // 如果文件不存在，添加表头
        if (!fs.existsSync(this.logFile)) {
            fs.writeFileSync(this.logFile, 
                "=".repeat(80) + "\n" +
                `价格审核日志 - ${new Date().toLocaleDateString()}\n` +
                "=".repeat(80) + "\n\n",
                'utf-8'
            );
        }
    }

    /**
     * 重置错误计数
     */
    private resetErrorCount(): void {
        this.errorCount = 0;
        logger.info("错误计数已重置");
    }

    /**
     * 增加错误计数并检查是否需要重启
     */
    private async incrementErrorCount(context: string): Promise<boolean> {
        this.errorCount++;
        logger.warn(`${context} 发生错误，当前错误计数: ${this.errorCount}/${this.MAX_ERROR_COUNT}`);

        if (this.errorCount >= this.MAX_ERROR_COUNT) {
            logger.error(`错误次数达到上限 ${this.MAX_ERROR_COUNT}，准备重启流程`);
            await this.restartReview();
            return true;
        }
        return false;
    }

    /**
     * 重启审核流程
     */
    private async restartReview(): Promise<void> {
        try {
            logger.info("开始重启审核流程...");
            this.resetErrorCount();
            await this.page.reload();
            await this.page.waitForLoadState('networkidle');
            await this.startReview();
        } catch (error) {
            logger.error(`重启审核流程失败: ${error}`);
            throw error;
        }
    }

    /**
     * 记录价格审核日志
     */
    private logRecord(sku: string, spu: string, minPrice: number | string, newPrice: number | string, action: string): void {
        try {
            const timestamp = new Date().toLocaleString();
            const logEntry = `[${timestamp}] 货号: ${sku}, SPU: ${spu}, 最低核价: ${minPrice}, 新申报价格: ${newPrice}, 操作: ${action}\n`;
            
            fs.appendFileSync(this.logFile, logEntry, 'utf-8');
            
            logger.info(`已记录日志: ${logEntry.trim()}`);
        } catch (error) {
            logger.error(`记录日志失败: ${error}`);
        }
    }

    /**
     * 打开价格审核页面
     */
    private async openPriceReviewPage(): Promise<void> {
        try {
            logger.info("正在打开上新主页面...");
            await this.page.goto("https://seller.kuajingmaihuo.com/main/product/seller-select?topTab=0&quickTab=1");
            await this.page.waitForTimeout(3000);
            logger.info("页面加载完成");
        } catch (error) {
            logger.error(`打开价格审核页面失败: ${error}`);
            throw error;
        }
    }

    /**
     * 检查并关闭弹窗
     */
    private async checkAndClosePopup(): Promise<boolean> {
        try {
            const popupTitle = await this.page.$('.PP_popoverTitle_5-114-0');
            
            if (popupTitle) {
                logger.info("发现弹窗，准备关闭...");
                
                const closeButton = await this.page.waitForSelector('.new-bell_container__eWEgQ', { state: 'visible' });
                await closeButton?.click();
                
                await this.page.waitForTimeout(1000);
                
                logger.info("弹窗已关闭");
                return true;
            }
            
            return false;
        } catch (error) {
            logger.error(`检查或关闭弹窗失败: ${error}`);
            return false;
        }
    }

    /**
     * 设置每页显示10条记录
     */
    private async setPageSize(): Promise<boolean> {
        try {
            logger.info("正在设置每页显示数量...");

            const pageSizeInput = await this.page.waitForSelector(
                '//*[@id="root"]/div/div/div/div[3]/div[2]/div[2]/ul/li[2]/div/div/div/div/div/div',
                { state: 'visible' }
            );
            await pageSizeInput?.click();
            
            // const option10 = await this.page.waitForSelector(
            //     '/html/body/div[6]/div/div/div/div/ul/li[1]',
            //     { state: 'visible' }
            // );
            // await option10?.click();

            const option10 = await this.page.waitForSelector(
                "li[role='option'] span:text('10')",
                { state: 'visible' }
            );
            await option10?.click();
            
            await this.page.waitForTimeout(2000);
            
            logger.info("已设置每页显示10条记录");
            return true;
        } catch (error) {
            logger.error(`设置每页显示数量失败: ${error}`);
            return false;
        }
    }

    /**
     * 获取分页信息
     */
    private async getPageInfo(): Promise<[number, number, number]> {
        try {
            // 先定位到分页组件
            const pagination = this.page.getByTestId('beast-core-pagination');
            await pagination.waitFor({ state: 'visible', timeout: 5000 });

            // 在分页组件内获取总条数
            const totalText = await pagination
                .locator('.PGT_totalText_5-117-0')
                .textContent() || '';
            const totalItems = parseInt(totalText.replace(/\D/g, '')) || 0;

            // 在分页组件内获取当前页码
            const currentPageText = await pagination
                .locator('.PGT_pagerItemActive_5-117-0')
                .textContent() || '1';
            const currentPage = parseInt(currentPageText);

            // 在分页组件内获取每页条数
            const pageSizeText = await pagination
                .locator('[data-testid="beast-core-select-htmlInput"]')
                .getAttribute('value') || '10';
            const pageSize = parseInt(pageSizeText);

            // 计算总页数
            const totalPages = Math.ceil(totalItems / pageSize);

            logger.info(`分页信息: 总条数 ${totalItems}, 当前页 ${currentPage}/${totalPages}, 每页 ${pageSize} 条`);

            return [currentPage, totalPages, pageSize];

        } catch (error) {
            logger.error(`获取分页信息失败: ${error}`);
            return [1, 100, 10]; // 返回默认值
        }
    }

    /**
     * 跳转到下一页
     */
    private async goToNextPage(): Promise<boolean> {
        try {
            const pagination = await this.page.waitForSelector(
                '//*[@id="root"]/div/div/div/div[3]/div[2]/div[2]/ul',
                { state: 'visible' }
            );
            
            try {
                const nextButton = await pagination?.$('.PGT_next_5-117-0');
                await nextButton?.click();
                logger.info("已跳转到下一页");
                return true;
            } catch (error) {
                logger.info("已到达最后一页，准备返回第一页");
                
                const firstPageButton = await this.page.waitForSelector(
                    '//*[@id="root"]/div/div/div/div[3]/div[2]/div[2]/ul/li[4]',
                    { state: 'visible' }
                );
                await firstPageButton?.click();
                
                await this.page.waitForTimeout(2000);
                logger.info("已返回第一页");
                return true;
            }
        } catch (error) {
            logger.error(`页面跳转失败: ${error}`);
            return false;
        }
    }

    /**
     * 获取商品信息
     */
    private async getProductInfo(rowElement: any): Promise<{
        spu: string;
        sku: string;
        code: string | undefined;
        minPrice: number | undefined;
        priceButton: any;
    }> {
        try {
            const spuElement = await rowElement.$('div.product-info_name__pXkCK');
            const spuInfo = await spuElement?.textContent() || "未找到SPU信息";

            const skuElement = await rowElement.$('div.use-columns_infoItem__iPDoh.use-columns_light__sKFcQ');
            const skuNumber = await skuElement?.textContent() || "未找到货号";

            const priceButton = await rowElement.$('button.BTN_outerWrapper_5-111-0');
            
            // 检查价格确认按钮是否存在
            if (!priceButton) {
                throw new Error(`未找到价格确认按钮 - SPU: ${spuInfo}, 货号: ${skuNumber}`);
            }
            
            const [code, minPrice] = this.priceMapper.getPriceInfo(skuNumber);
            
            logger.info(`\n商品信息:
SPU: ${spuInfo}
货号: ${skuNumber}
识别码: ${code}
最低核价: ${minPrice}
找到价格确认按钮: ${await priceButton?.textContent()}`);

            return {
                spu: spuInfo,
                sku: skuNumber,
                code,
                minPrice,
                priceButton
            };
        } catch (error) {
            logger.error(`获取商品信息失败: ${error}`);
            throw error;
        }
    }

    /**
     * 处理价格确认弹窗
     */
    private async handlePriceConfirmation(minPrice: number, sku: string, spu: string): Promise<boolean> {
        try {
            logger.info("\n开始处理价格确认弹窗...");

            const modal = await this.page.waitForSelector(".MDL_innerWrapper_5-111-0", { state: 'visible' });
            
            const priceElement = await this.page.waitForSelector(
                ".TB_whiteTr_5-111-0 td:nth-child(5) span:nth-child(2)",
                { state: 'visible' }
            );
            const newPrice = parseFloat(await priceElement?.textContent() || '0');
            
            logger.info(`最低核价: ${minPrice}`);
            logger.info(`新申报价格: ${newPrice}`);

            if (newPrice > minPrice) {
                logger.info("新申报价格大于最低核价，直接确认");
                
                const confirmButton = await this.page.waitForSelector(
                    ".MDL_okBtn_5-111-0",
                    { state: 'visible' }
                );

                await this.page.waitForTimeout(500);
                await confirmButton?.click();
                
                this.logRecord(sku, spu, minPrice, newPrice, "确认提交");
            } else {
                logger.info("新申报价格小于等于最低核价，选择放弃调整报价");
                
                const dropdown = await this.page.waitForSelector(
                    "#operatorType .ST_head_5-111-0",
                    { state: 'visible' }
                );
                await dropdown?.click();
                
                const abandonOption = await this.page.waitForSelector(
                    "li[role='option'] span:text('放弃调整报价')",
                    { state: 'visible' }
                );
                await abandonOption?.click();

                const confirmButton = await this.page.waitForSelector(
                    ".MDL_okBtn_5-111-0",
                    { state: 'visible' }
                );

                await this.page.waitForTimeout(500);
                await confirmButton?.click();
                
                this.logRecord(sku, spu, minPrice, newPrice, "放弃调整报价");
            }

            await this.page.waitForTimeout(2000);
            logger.info("价格确认处理完成");
            
            return true;
        } catch (error) {
            logger.error(`处理价格确认弹窗失败: ${error}`);
            this.logRecord(sku, spu, minPrice, "获取失败", `处理失败: ${error}`);
            return false;
        }
    }

    /**
     * 滚动到元素位置
     */
    private async scrollToElement(element: any): Promise<void> {
        try {
            await this.page.evaluate((el) => {
                el.scrollIntoView({ behavior: 'smooth', block: 'center' });
            }, element);
            await this.page.waitForTimeout(200);
        } catch (error) {
            logger.error(`滚动到元素位置失败: ${error}`);
        }
    }

    /**
     * 处理商品列表
     */
    private async processProductList(): Promise<void> {
        try {
            const [currentPage, totalPages, itemsPerPage] = await this.getPageInfo();
            logger.info(`当前页: ${currentPage}, 总页数: ${totalPages}, 每页条数: ${itemsPerPage}`);
            
            let consecutiveErrors = 0;
            while (true) {
                const productList = await this.page.waitForSelector(
                    '#root > div > div > div > div.TB_outerWrapper_5-117-0.TB_bordered_5-117-0.TB_notTreeStriped_5-117-0 > div.TB_inner_5-117-0 > div > div.TB_body_5-117-0 > div > div > table > tbody',
                    { state: 'visible' }
                );
                const rows = await productList?.$$('tr') || [];

                for (let index = 0; index < rows.length; index++) {
                    try {
                        logger.info(`\n处理第 ${index + 1} 个商品:`);
                        
                        await this.scrollToElement(rows[index]);
                        
                        const productInfo = await this.getProductInfo(rows[index]);
                        
                        await productInfo.priceButton?.click();
                        
                        if (productInfo.minPrice !== undefined) {
                            await this.handlePriceConfirmation(
                                productInfo.minPrice,
                                productInfo.sku,
                                productInfo.spu
                            );
                        }
                        
                        logger.info(`已处理第 ${index + 1} 个商品`);
                        consecutiveErrors = 0; // 重置连续错误计数
                        
                        await this.page.waitForTimeout(2000);
                        
                    } catch (error) {
                        logger.error(`处理第 ${index + 1} 个商品时发生错误: ${error}`);
                        this.logRecord(
                            `第${index + 1}个商品`,
                            "获取失败",
                            "获取失败",
                            "获取失败",
                            `处理失败: ${error}`
                        );
                        
                        consecutiveErrors++;
                        if (consecutiveErrors >= 5) {
                            logger.error(`连续处理 ${consecutiveErrors} 个商品失败，准备重启流程`);
                            await this.restartReview();
                            return;
                        }
                        continue;
                    }
                }

                const [currentPage, totalPages, itemsPerPage] = await this.getPageInfo();

                if (currentPage < totalPages) {
                    if (!await this.goToNextPage()) {
                        logger.info("无法跳转到下一页，处理结束");
                        break;
                    }
                    await this.page.waitForTimeout(2000);
                } else {
                    logger.info("已处理完所有页面，准备重新从第一页开始检查");
                    
                    const firstPageButton = await this.page.waitForSelector(
                        '//*[@id="root"]/div/div/div/div[3]/div[2]/div[2]/ul/li[4]',
                        { state: 'visible' }
                    );
                    await firstPageButton?.click();
                    
                    await this.page.waitForTimeout(2000);
                    
                    logger.info("已返回第一页，开始第二轮检查");
                    continue;
                }
            }
        } catch (error) {
            logger.error(`处理商品列表失败: ${error}`);
            if (await this.incrementErrorCount('处理商品列表')) {
                return;
            }
            throw error;
        }
    }

    /**
     * 开始价格审核流程
     */
    public async startReview(): Promise<void> {
        try {
            await this.openPriceReviewPage();
            await this.setPageSize();
            await this.processProductList();
            logger.info("价格审核流程初始化完成");
            this.resetErrorCount(); // 成功完成后重置错误计数
        } catch (error) {
            logger.error(`价格审核流程初始化失败: ${error}`);
            if (await this.incrementErrorCount('价格审核流程初始化')) {
                return;
            }
            throw error;
        }
    }
} 